src/Psalm/Config.php

```diff
@@ -1861,8 +1873,42 @@ public function addStubFile(string $stub_file)
         $this->stub_files[] = $stub_file;
     }
 
+    public function getPhpVersion(): ?string
+    {
+        if (isset($this->configured_php_version)) {
+            return $this->configured_php_version;
+        }
+
+        return $this->getPHPVersionFromComposerJson();
+    }
+
     private function setBooleanAttribute(string $name, bool $value): void
     {
         $this->$name = $value;
     }
+
+    /**
+     * @psalm-suppress MixedAssignment
+     * @psalm-suppress MixedArrayAccess
+     */
+    private function getPHPVersionFromComposerJson(): ?string
+    {
+        $composer_json_path = $this->base_dir . DIRECTORY_SEPARATOR. 'composer.json';
+
+        if (file_exists($composer_json_path)) {
+            if (!$composer_json = json_decode(file_get_contents($composer_json_path), true)) {
+                throw new \UnexpectedValueException('Invalid composer.json at ' . $composer_json_path);
+            }
+            $php_version = $composer_json['require']['php'] ?? null;
+
+            if ($php_version) {
+                foreach (['5.4', '5.5', '5.6', '7.0', '7.1', '7.2', '7.3', '7.4', '8.0'] as $candidate) {
```

Comment:
I don't like having to do this loop here hard-coding versions, but IMHO there's no way to deterministically reverse a composer range to end up with a minimum version.

What still is a problem is that the knowledge of what versions of PHP psalm has actual support for isn't centralized but so far only existed in this regex here:

https://github.com/vimeo/psalm/blob/5d15ef062e0efdf7d3d820311148f4816943d176/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php#L1164

maybe for the future it would be worth to actually have proper API to get to this information so further duplication isn't needed.